src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -11,32 +11,62 @@ namespace System.Net
 {
     internal class IPAddressParser
     {
+
+        internal static unsafe bool ContainsColon(char* ipStringPtr, int ipStringPtrLength)
+        {
+            for (var i = 0; i < ipStringPtrLength; i++)
+            {
+                if (ipStringPtr[i] == ':')
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
         internal static unsafe IPAddress Parse(string ipString, bool tryParse)
         {
-            if (ipString == null)
+            fixed (char* ipStringPtr = ipString)
+            {
+                return Parse(ipStringPtr, ipString.Length, tryParse);
+            }
+        }
+
+        internal static unsafe IPAddress Parse(ReadOnlySpan<char> ipSpan, bool tryParse)
+        {
+            fixed (char* ipStringPtr = &ipSpan.DangerousGetPinnableReference())
+            {
+                return Parse(ipStringPtr, ipSpan.Length, tryParse);
+            }
+
+        }
+   
+        internal static unsafe IPAddress Parse(char* ipStringPtr, int ipStringPtrLength, bool tryParse)
```

Comment:
Could this be written in terms of `ReadOnlySpan<char>` rather than in terms of `char*`?  That'd a) be safer and b) avoid pinning for unnecessary periods of time.  Regardless, the null check below can be moved up into the string-based overload.